### ISSUE TITLE
score fuzzy term by distance

### ISSUE THREAD

**Issue by buster** [quickwit-oss/tantivy#563]
When playing around with tantivy, i came up with a query including multiple fuzzy term queries.
Unvortunately the results don't make much sense (e.g. imagine a text field "the quick brown fox" and some other completely different text fields and then do a boolean query with multiple fuzzy terms. The results are completely unusable, because even an exact search term results in some completely different term scored the same).
I think it comes down to this: The single fuzzy term query doesn't make a difference between exact matches or some levenshtein distance.
The test should find the second added document first (with a higher score), but doesn't.

```
#[cfg(test)]
mod test {
    use tantivy::query::FuzzyTermQuery;
    use tantivy::collector::TopDocs;
    use tantivy::schema::Schema;
    use tantivy::schema::{TEXT, STORED};
    use tantivy::Index;
    use tantivy::Term;

    #[test]
    pub fn test_fuzzy_term() {
        let mut schema_builder = Schema::builder();
        let country_field = schema_builder.add_text_field("country", TEXT | STORED);
        let schema = schema_builder.build();
        let index = Index::create_in_ram(schema);
        {
            let mut index_writer = index.writer_with_num_threads(1, 10_000_000).unwrap();
            index_writer.add_document(doc!(
                country_field => "WENN ROT WIE RUBIN",
            ));
            index_writer.add_document(doc!(
                country_field => "WENN ROT WIE ROBIN",
            ));
            index_writer.commit().unwrap();
        }
        let reader = index.reader().unwrap();
        let searcher = reader.searcher();
        {
            let term = Term::from_field_text(country_field, "robin");

            let fuzzy_query = FuzzyTermQuery::new(term, 2, true);
            let top_docs = searcher
                .search(&fuzzy_query, &TopDocs::with_limit(100))
                .unwrap();
            assert_eq!(top_docs.len(), 2, "Expected 2 documents");
            let (score, adr) = top_docs[0];
            let document = index.schema().to_named_doc(&searcher.doc(adr).expect("document"));
            let json = index.schema().to_json(&searcher.doc(adr).expect("document"));
            println!("{}", json);
            println!("{:?}", top_docs);
            assert_eq!(document.0.get("country").expect("second document")[0].text().unwrap(), "WENN ROT WIE ROBIN");
        }
    }
}
```

Somehow the distance needs to be considered in the scoring, else the fuzzy query doesn't make much sense.

Or is there some other way to achieve this, i didn't see?

**Comment by fulmicoton** [comment]
You are absolutely correct.

So a simple solution would be to change the score of the fuzzytermquery depending on the distance it hit. 

Bad news it was never implemented because the fst automaton did not expose the automaton state when streaming over matching terms.
Good news, partly for this reason I forked the fst crate and someone already wrote a PR that has never been imported in the fst crate. 

We simply need to merge it in tantivy-fst and tweak the score in the fuzzytermquery to solve your issue. Are you interested in the ticket?

**Comment by buster** [comment]
Yes, i'm interested in this :)

**Comment by fulmicoton** [comment]
Rephrasing. Are you interesting in addressing this ticket? Do you need pointers to the stuff I was talking about above?

**Comment by buster** [comment]
I didn't understand that,but if you can point out the relevant pieces, I'd give it a try.

**Comment by fulmicoton** [comment]
no problem I can search for volunteers on the chat :)

**Comment by buster** [comment]
Just wondering: How is this going? Is there a branch for this feature? Also, where is the fst crate fork and the PR?

**Comment by hntd187** [comment]
The PR for fst was since deleted, I can't find the commit anywhere, so it'll have to ported over by hand. This is gonna take some time. I probably won't have much time to look at this in the next few days, but I'll take to the grind a bit later in the week.

**Comment by hntd187** [comment]
@fulmicoton how do we want to tackle this PR for fst? I don't see much way other than implementing it by hand from that PR, which I had to dig up, but the commit itself is gone.

**Comment by fulmicoton** [comment]
@hntd187  Can you ask his author? He is nice and usually replies really fast.

Also if you do not rewrite this code yourself, make sure that there is no license conflict and he gets the credit somewhere and his happy with the visibility of this credit.

**Comment by trinity-1686a** [comment]
@hntd187 is it pr 61? If so, you can get the commit back by running `git fetch origin +refs/pull/61/head && git checkout FETCH_HEAD`

**Comment by hntd187** [comment]
So @fulmicoton we got this in fst now, we're all ready to hook it up in tantivy right? I'll get that started if you have any more thoughts from out gitter convo

**Comment by lerouxrgd** [comment]
Could this be solved by leveraging `StreamWithState` from `fst` crate which is now available (details in [this commit](https://github.com/BurntSushi/fst/commit/fc757c65467d8033019de37c990b302bfbf43276)) ?

**Comment by lerouxrgd** [comment]
If my understanding is correct, here are the things to do:

- Add a `search_with_state` method to `tantivy_fst::Map` like [fst::Map::search_with_state](https://docs.rs/fst/0.4.5/fst/struct.Map.html#method.search_with_state) in `fst`.

- Add a `TermWithStateStreamerBuilder` (similar to [TermStreamerBuilder](https://docs.rs/tantivy/0.14.0/src/tantivy/termdict/fst_termdict/streamer.rs.html#13-73)) to `tantivy::termdict::fst_termdict::streamer` but that wraps a [tantivy_fst::map::StreamWithStateBuilder](https://docs.rs/tantivy-fst/0.3.0/tantivy_fst/map/struct.StreamWithStateBuilder.html) instead of a [tantivy_fst::map::StreamBuilder](https://docs.rs/tantivy-fst/0.3.0/tantivy_fst/map/struct.StreamBuilder.html), with functionalities similar to [fst::map::StreamWithStateBuilder](https://docs.rs/fst/0.4.5/fst/map/struct.StreamWithStateBuilder.html) in `fst`. 

- Add a `TermWithStateStreamer` (similar to [TermStreamer](https://docs.rs/tantivy/0.14.0/src/tantivy/termdict/fst_termdict/streamer.rs.html#77-151)) to `tantivy::termdict::fst_termdict::streamer` but that wraps a [tantivy_fst::map::StreamWithState](https://docs.rs/tantivy-fst/0.3.0/tantivy_fst/map/struct.StreamWithState.html) instead of a [tantivy_fst::map::Stream](https://docs.rs/tantivy-fst/0.3.0/tantivy_fst/map/struct.Stream.html).

- Add a method `search_with_state` to `tantivy::fst_termdict::termdict::TermDictionary` that returns a `TermWithStateStreamerBuilder`.

- Add a private method `automaton_stream_with_state` to [tantivy::query::AutomatonWeight](https://docs.rs/tantivy/0.14.0/tantivy/query/struct.AutomatonWeight.html) (similar to [automaton_stream](https://docs.rs/tantivy/0.14.0/src/tantivy/query/automaton_weight.rs.html#33-40)) but that leverages `TermDictionary::search_with_state` to return a `TermWithStateStreamer` through `TermWithStateStreamerBuilder::into_stream`.

- Use  `automaton_stream_with_state` instead of `automaton_stream` in [AutomatonWeight::scorer](https://docs.rs/tantivy/0.14.0/src/tantivy/query/automaton_weight.rs.html#53).

- Retrieve the now available [A::State](https://docs.rs/tantivy-fst/0.3.0/src/tantivy_fst/map.rs.html#1067-1078) in  `TermWithStateStreamer::advance` (similar to [TermStreamer::advance](https://docs.rs/tantivy/0.14.0/src/tantivy/termdict/fst_termdict/streamer.rs.html#96)) and return it in addition to `TermInfo` in `TermWithStateStreamer::value` (similar to [TermStreamer::value](https://docs.rs/tantivy/0.14.0/src/tantivy/termdict/fst_termdict/streamer.rs.html#138-140)).

- Make use of `A::State` returned by `TermWithStateStreamer::value` in [AutomatonWeight::scorer](https://docs.rs/tantivy/0.14.0/src/tantivy/query/automaton_weight.rs.html#48-72) to score the documents of the `doc_bitset` retrieved with `term_info`. Finally return an appropriate `Scorer` implementation (probably different from the current `ConstScorer`).

That last step isn't completely clear for me, however do these steps look good ?

**Comment by fulmicoton** [comment]
@lerouxrgd This is a perfect summary of what needs to done.

The last step is not too difficult. The state can be converted into a levenshtein distance.
https://github.com/tantivy-search/levenshtein-automata/blob/master/src/dfa.rs#L70

Scoring given `(overall number of  matching docs, docfreq matching term, levenshtein_distance)` is then a non-trivial problem.

**Comment by maufl** [comment]
@lerouxrgd Are you planning to work on it? I was thinking on giving it a try when I understand everything.

What I'm currently not understanding is, the `Scorer` trait represents a set of documents (as it can be downcast to a `DocSet`? This part was difficult to follow in the code and documentation), but the `score` method returns only a single `Score`? It seems to me, documents matched by the same `Query -> Weight -> Scorer` are supposed to have the same score? You could of course return a different score, depending on the current `doc()` of the `DocSet`, but that is probably not very intuitive.

**Comment by fulmicoton** [comment]
@maufl Generally the scorer is kind of an iterator, on the sorted document set. Due to the strange nature of queries based on the automatonweight, the docset is actually entirely precalculated and stored into a bitset, and this is the iterator that is returned.

The iterator is not a rust Iterator...
You are meant to call advance to advance it, and then .doc() and .score() respectively give you the doc id and the score of the current document.

**Comment by maufl** [comment]
@fulmicoton Understood, then I think my intuition was right.

@lerouxrgd I think instead of implementing `search_with_state` in `fst` we can use `StreamBuilder::with_state` as implemented by @hntd187. We could extend the `TermStreamer` to take a `StreamWithState`and hold a `current_state: A::State`. This would require much less changes. The only downside is, that the `TermStreamer` will then always make a clone of the state.

**Comment by lerouxrgd** [comment]
@maufl Yes good point, it might be better to use `StreamBuilder::with_state` instead. I guess I can give all this a try, at least until the last step I listed where the real calculation happens. I am also a bit confused by the `DocSet` and `Scorer` interactions, I would need to dig into [TermWeight::specialized_scorer](https://docs.rs/tantivy/0.14.0/src/tantivy/query/term_query/term_weight.rs.html#102-130) and [TermScorer](https://docs.rs/tantivy/0.14.0/src/tantivy/query/term_query/term_scorer.rs.html#103-127) to understand how to score documents of a result set differently.

@fulmicoton Thank you for confirming this is correct. Besides, what would prevent regular BM25 scoring in `AutomatonWeight::scorer` instead of `ConstScorer` with 1.0 ? Do we have enough info at this moment to calculate it through `BM25Weight::for_one_term` for instance ?

**Comment by lerouxrgd** [comment]
@maufl FYI I implemented the steps I listed that bubble up `Automaton::State` in `AutomatonWeight::scorer` in [this branch](https://github.com/lerouxrgd/tantivy/tree/fst-state) following your advice. Now I guess the question is, how to score documents.

**Comment by maufl** [comment]
@lerouxrgd That looks very good. I was thinking about the scoring and scorer. What about calculating the score as `1 / (1 + distance)`? For a perfect match we would have score `1` and for a greater distance a smaller score. Given that we will probably always have only very little different values for the distance/score, I thought about writing a new scorer that wraps multiple const scorers, one for each distance/score. The new scorer could also directly contain a vector of `(BitSetDocSet, Score)`.

**Comment by lerouxrgd** [comment]
@maufl Yes, I was thinking of the same score. Actually the tricky part is still to get the automaton state because all we have in the API is `A: Automaton` and it is impossible to simply get a the Levenshtein distance from `A::State` (because we don't know what this type is at compile time). The only workaround I found is to use reflection at runtime. I am preparing a PR with all this.

**Comment by maufl** [comment]
@lerouxrgd Alternatively you can implement `Weight` for both specializations of `AutomatonWeight` seperately. I.e. `impl Weight for AutomatonWeight<Regex>` and `impl Weight for AutomatonWeight<DFAWrapper>`. That way you will have a concrete automata type and can use the `distance` method directly.

**Comment by lerouxrgd** [comment]
Oh yes good idea, let's see what the maintainers think is best.

**Comment by maufl** [comment]
I'm also unsure whether having an extra `TermWithStateStreamerBuilder` and `TermWithStateStreamer` is better than just merging the `current_state` field into `TermStreamer`. It seems like a lot of duplicated code to my. Let's wait and see what @fulmicoton says.

**Comment by fulmicoton** [comment]
Overall the PR looks good except one very important detail, I commented on the PR. Good job! The hardest part is done...

**Comment by fulmicoton** [comment]
@lerouxrgd @maufl  Following on the PR. As we are discovering scoring really is tricky...

The more I think about it, the more I feel incorporating the lev distance with a score is hard. The best I can think for a given query term score, is to pick the minimum distance observe and take the max BM25 score for the term at this distance.

In many ways, a more satisfying to implement fuzzy search would be to use a algolia like approach.
The scoring would like as a lexicographical order of different scoring methods.

First: Do all of the terms match perfectly? If not enough doc are matching, or there is a tie, then try with 
an overall levenshtein distance of 1 (sum of the lev distance of all terms).
Then try with an overall levenshtein distance of 2.
etc.
In case of a tie, break the tie with BM25. 

It makes sense both as great trade off between search quality and usability, and for performance reasons.
The idea there would be to have a search routine independent from the usual tantivy paradigm of query pushing docs to  collector.

The routine would just be in charge of computing a TopK.

```
fn fuzzy_search(query: &dyn Query, k: usize /* possibly one extra argument called ranking_policy */) -> Vec<DocAddress> {
   /// ...
}
```

The routine would work actually work like the policy above. Instead of trying to score all of the docs with a levenshtein 2 and pay a massive price when it is most likely not useful, try distance 0 first... and then increment from there.

In algolia, there are more criteria than just lev distance and bm25, and their order can be configured with a Policy.


---- 

Do you have projects who would benefit from this PR ? @lerouxrgd @maufl In its current shape, I suspect the approach in the PR will be not be useful to anyone and add to tantivy's tech debt.
If you do not have any use for it, I think we should drop this PR or try to extract from the PR something that could be useful for users.

(Apologies for not seeing this coming earlier)

**Comment by maufl** [comment]
I have a project that is a product search for local shops. I want to build a database with the offers of the shops in my city and have a full text search for it. The search should allow fuzzy searching, i.e. return results even if there is a small typo, but I want exact matches to appear first.

**Comment by fulmicoton** [comment]
@maufl sweet. Do you have an idea of what is the correct order for these results? Do you want to see a result count and facetting, and how do you define that with a fuzzy query?

**Comment by maufl** [comment]
@fulmicoton (Disclaimer, I don't know much about full text search in general) I'd like to sort the results based on how good the match was, i.e. whether it was an exact match or not, and maybe on the number or matches in one document. Furthermore, I might like to give matches in a certain field a higher weight, i.e. a match in "name" is better than in "description". However, I did not try to actually implement this yet.
I also want to see the result count and maybe do pagination. I don't plan to do facetting, as I don't think there will be enough data for it. (I think I will have a name and hopefully a description for each product but likely not much more).

Right now I'm building a query like this:
```rust
    fn make_query(&self, query_string: &str) -> Box<dyn Query> {
        let mut parts: Vec<(Occur, Box<dyn Query>)> = Vec::new();
        let fields = vec![
            self.name_field,
            self.description_field,
            self.category_field,
            self.color_field,
        ];
        for search_term in query_string.to_ascii_lowercase().split(" ") {
            if search_term.is_empty() {
                continue;
            }
            for field in fields.iter() {
                parts.push((
                    Occur::Should,
                    Box::new(FuzzyTermQuery::new_prefix(
                        Term::from_field_text(*field, search_term),
                        1,
                        true,
                    )),
                ))
            }
        }
        Box::new(BooleanQuery::from(parts))
    }
```
Then I use the `Count` and `TopDocs` collector with limit and offset.
I hope I did not (ab-)use tantivy in a completely wrong way here :D

**Comment by lerouxrgd** [comment]
@fulmicoton I am not using it directly now, so as far as I am concerned it is fine to discard this PR. I see that correct scoring is indeed tricky, however I still feel that just ordering results with `1 / (1 + dist)` should be doable (and fixes this original issue) even if it is not the best possible score.

**Comment by afbarbaro** [comment]
Is there a path forward for this issue? I have a use case that really needs fuzzy search on multiple terms to handle user typos and this issue is a blocker.
Thanks!

**Comment by overthetop** [comment]
Anything new on this issue? I saw that this was fixed in paradedb fork of tantivy https://github.com/paradedb/tantivy/pull/1/files. Can we just make this fix in here as well? @fulmicoton

**Comment by Boscop** [comment]
Yes, it would be great if ParadeDB's fix can be integrated here 🙏
My use case requires fuzzy search to handle typos & transposed letters in indexed documents.

**Comment by mxfunc** [comment]
@fulmicoton I would like to work on this. I am trying to understand the comment where you have mentioned having a separate search path - `fuzzy_search`. I looked into code to understand a few roadblocks and if you can help with your preferred approach, I can take a crack at writing a spec for Algolia-like ranking that you mentioned.


Because you mentioned in your comment that fuzzy_search takes a `dyn`, I assume that you want the fuzzy queries to support boolean queries as well (which Algolia does not, btw). This makes the accumulation logic non-trivial. The current Query AST execution is very tightly coupled with scorer and weights.

- I see that right now we only have one type of scorer per term query. If we want BM25 to be a tie-breaker, we either need to scorer return type to change or the ability to calculate multiple scores.
- The accumulation logic needs to be modified as well for boolean and other queries as we have to accumualate a score of vectors.

One approach I explored was to create a parallel Ranker AST from existing Query AST, use the existing Query AST for matches but then runs every document through Ranker AST. And we do multiple passes for typo length. My inuition is that this is going to very expensive as the number of matches can be very large. This gymanistics feels weird because I didn't want to make changes to existing data structurues.

Do you have any guidance as to what the right approach could be and what trade-offs we could make to simplify this?

**Comment by rafael-patronilo** [comment]
Hi, I see the discussion revolves a lot about how to score fuzzy matches, I wanted to ask if you have considered using similarity:

`1 - ( dist(a, b) / max(len(a), len(b)) )`

Intuitively this formula gives "The proportion of matching characters between both strings". This results in a value between 0 and 1 where 1 means exact match and 0 means no match at all. It balances distances between small strings, where a difference of 1 character is much more meaningful than for a large string.

Currently, I am just doing the sort of iterative approach you suggested where I try larger and larger distances until I get a match. I did this without changing tantivy code, I just run multiple searches with modified queries. This works fine in my use case.
